client_test/crash_test.go

```diff
@@ -0,0 +1,59 @@
+// Package client_test contains helpers and integration tests for go-nitro clients
+package client_test // import "github.com/statechannels/go-nitro/client_test"
+
+import (
+	"testing"
+
+	"github.com/statechannels/go-nitro/client"
+	"github.com/statechannels/go-nitro/client/engine"
+	"github.com/statechannels/go-nitro/client/engine/chainservice"
+	"github.com/statechannels/go-nitro/client/engine/messageservice"
+	"github.com/statechannels/go-nitro/client/engine/store"
+	"github.com/statechannels/go-nitro/types"
+)
+
+func TestCrashTolerance(t *testing.T) {
+
+	// Setup logging
+	logFile := "test_crash_tolerance.log"
+	truncateLog(logFile)
+	logDestination := newLogWriter(logFile)
+
+	// Setup chain service
+	sim, bindings, ethAccounts, err := chainservice.SetupSimulatedBackend(3)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	chainA, err := chainservice.NewSimulatedBackendChainService(sim, bindings, ethAccounts[0], logDestination)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	chainB, err := chainservice.NewSimulatedBackendChainService(sim, bindings, ethAccounts[2], logDestination)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// End chain service setup
+
+	broker := messageservice.NewBroker()
+
+	// Client setup
+	storeA := store.NewMemStore(alice.PrivateKey)
+	messageserviceA := messageservice.NewTestMessageService(alice.Address(), broker, 0)
+	clientA := client.New(messageserviceA, chainA, storeA, logDestination, &engine.PermissivePolicy{}, nil)
+
+	clientB, _ := setupClient(bob.PrivateKey, chainB, broker, logDestination, 0)
+	// End Client setup
+
+	// test successful condition for setup / teadown of unused ledger channel
+	{
+		channelId := directlyFundALedgerChannel(t, clientA, clientB, types.Address{})
+
+		anotherClientA := client.New(messageserviceA, chainA, storeA, logDestination, &engine.PermissivePolicy{}, nil)
```

Comment:
Here we spin up a new client with the same messaging service, chain service and store, and use that for the rest of the test. If the test fails, it's a sign that something important exists only in memory outside of the store.

The test currently fails.

Reply:
I believe the test failed because the old client continued to consume events from the chain service, starving the new client of that information. 

Stopping the old client before starting the new one makes the test pass [1733da1](https://github.com/statechannels/go-nitro/pull/1138/commits/1733da120a161475229fefdcbe97839111d0ad62) 🎉